nearcore/src/runtime/mod.rs

```diff
@@ -716,9 +716,11 @@ impl RuntimeAdapter for NightshadeRuntime {
         self.tries.clone()
     }
 
+    // TODO (#7327): Make usage of flat state conditional on prev_hash
     fn get_trie_for_shard(&self, shard_id: ShardId, prev_hash: &CryptoHash) -> Result<Trie, Error> {
         let shard_uid = self.get_shard_uid_from_prev_hash(shard_id, prev_hash)?;
-        Ok(self.tries.get_trie_for_shard(shard_uid))
+        let use_flat_state = cfg!(feature = "protocol_feature_flat_state");
```

Comment:
What's the actual logic we want to use to enable flat_store? I would like to avoid every call-site passing an extra parameter, and somehow design the API such that it just works. 

I also wonder if `use_flat_state`  should be a parameter of the trie at all? The way I understand it, trie is a persistent (functional)  data structure, which stores all historical evolution of the state. To get a particular snapshot, you need a trie and a state root. 

If flat_state is what we store for the tip, then "use_flat_state" seems to be a property of a particular state-root, rather than a property of the trie.

Reply:
In the protocol logic, Trie represents the state, e.g. we give trie (state) keys to it and ask for values. Flat state has the same purpose, and serves as trie optimization for roots close to the tip. That's why I think they should live in the same structure, possibly I need to rename `Trie` to `State`, `TrieUpdate` to `StateUpdate`, etc.

As for `use_flat_state`, I see two ways to handle decision whether to use flat state or not:
* On `Trie` creation - flat state struct will be either `None` or `Some(FlatState)`
* On each `get` operation - make decision based on `state_root`

I prefer the first one, because once Trie is created, we always query it with the same root, and I don't see future use-cases when it won't be the case. Moreover, I think that `root` has to become a field of the `Trie` and it should be removed from `get` and `get_ref` parameters. So if we know a root, we also know if we use flat state or not.

Reply:
>Moreover, I think that root has to become a field of the Trie and it should be removed from get and get_ref parameters. 

Yeah, this clear my concern! If we "bind" trie to the specific state root, the API makes perfect sense!